File: apps/Camera-Vitals/Android/app/src/main/java/ai/zetic/demo/cameravitals/state/MeasurementState.kt
package ai.zetic.demo.cameravitals.state

/** Top-level finite-state machine that drives which screen is shown. */
sealed interface MeasurementState {
    data class LoadingModel(val progress: Float) : MeasurementState
    data object PermissionDenied : MeasurementState
    data class Warmup(val framesFilled: Int) : MeasurementState
    data object Live : MeasurementState
    data class ErrorState(val message: String) : MeasurementState
}

/** Result of a guided 30-second measurement. */
data class MeasurementReport(
    val avgBPM: Int,
    val minBPM: Int,
    val maxBPM: Int,
    val avgQuality: Double,
    val series: List<Double>
)

File: apps/Camera-Vitals/iOS/CameraVitals/State/MeasurementState.swift
import Foundation

/// Top-level finite-state machine that drives which screen is shown.
enum MeasurementState: Equatable {
    case loadingModel(progress: Float)
    case permissionDenied
    case warmup(framesFilled: Int)   // buffer filling for the first time
    case live                        // continuous readout available
    case error(message: String)
}

/// Result of a guided 30-second measurement.
/// `Identifiable` with a stable per-instance `id` so the report sheet keeps a constant identity
/// across the frequent live-update re-renders behind it (otherwise the sheet flickers).
struct MeasurementReport: Equatable, Identifiable {
    let id = UUID()
    let avgBPM: Int
    let minBPM: Int
    let maxBPM: Int
    let avgQuality: Double
    let series: [Double]
}
